### PR TITLE
Implement NooBaa and test NooBaa bucket creation

### PR DESCRIPTION
**Pull request by MeridianExplorer**

### CODE REVIEW

File: ocs_ci/utility/noobaa.py [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+        )
+        creds_secret_name = (
+            results.get('items')[0].get('status').get('accounts')
+            .get('admin').get('secretRef').get('name')
+        )
+        ocp_obj2 = OCP(kind='secret', namespace='noobaa')
+        results2 = ocp_obj2.get(creds_secret_name)
+
+        noobaa_access_key = base64.b64decode(results2.get('data').get('AWS_ACCESS_KEY_ID')).decode('utf-8')
+        noobaa_secret_key = base64.b64decode(results2.get('data').get('AWS_SECRET_ACCESS_KEY')).decode('utf-8')
```
Comment (ebenahar) [inline ocs_ci/utility/noobaa.py]: line too long

File: ocs_ci/utility/noobaa.py [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+        results = ocp_obj.get()
+        endpoint = (
+            results.get('items')[0].get('status').get('services')
+            .get('serviceS3').get('externalDNS')[0]
+        )
+        creds_secret_name = (
+            results.get('items')[0].get('status').get('accounts')
+            .get('admin').get('secretRef').get('name')
+        )
+        ocp_obj2 = OCP(kind='secret', namespace='noobaa')
```
Comment (ebenahar) [inline ocs_ci/utility/noobaa.py]: call it secret_ocp_obj

File: tests/manage/noobaa/test_create_bucket.py [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+from ocs_ci.framework.pytest_customization.marks import tier1
+from ocs_ci.utility import noobaa
+
+
+logger = logging.getLogger(__name__)
+
+
+@pytest.fixture()
+def noobaa_obj():
+    """
```
Comment (ebenahar) [inline tests/manage/noobaa/test_create_bucket.py]: add description

File: ocs_ci/utility/noobaa.py [hunk 4]
```diff
@@ -0,0 +1,92 @@
+import base64
```
Comment (ebenahar) [inline ocs_ci/utility/noobaa.py]: As this is an OCS related resource, please move the file under ocs_ci/ocs/resources

File: tests/manage/noobaa/test_create_bucket.py [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+    @pytest.mark.skipif(
+        condition=config.ENV_DATA['platform'] != 'AWS',
+        reason="Tests are not running on AWS deployed cluster"
+    )
+    def test_s3_bucket_creation(self,  _obj, buckets):
+        """
+        Test bucket creation using the S3 SDK
+        """
+
+        bucketname = create_unique_resource_name(self.__class__.__name__.lower(), 's3-bucket')
```
Comment (ebenahar) [inline tests/manage/noobaa/test_create_bucket.py]: Please add logging

File: ocs_ci/ocs/resources/noobaa.py [hunk 6]
```diff
@@ -29,11 +29,13 @@ def __init__(self):
...
             .get('admin').get('secretRef').get('name')
         )
-        ocp_obj2 = OCP(kind='secret', namespace='noobaa')
-        results2 = ocp_obj2.get(creds_secret_name)
+        secret_ocp_obj = OCP(kind='secret', namespace='noobaa')
+        results2 = secret_ocp_obj.get(creds_secret_name)
 
-        noobaa_access_key = base64.b64decode(results2.get('data').get('AWS_ACCESS_KEY_ID')).decode('utf-8')
-        noobaa_secret_key = base64.b64decode(results2.get('data').get('AWS_SECRET_ACCESS_KEY')).decode('utf-8')
+        noobaa_access_key = base64.b64decode(
```
Comment (ebenahar) [inline ocs_ci/ocs/resources/noobaa.py]: Should be:

```python
        noobaa_access_key = base64.b64decode(
        noobaa_secret_key = base64.b64decode(results2.get('data').get('AWS_SECRET_ACCESS_KEY')).decode('utf-8')	            results2.get('data').get('AWS_ACCESS_KEY_ID')
).decode('utf-8')
```

File: ocs_ci/ocs/resources/noobaa.py [hunk 7]
```diff
@@ -29,11 +29,13 @@ def __init__(self):
...
+            results2.get('data').get('AWS_ACCESS_KEY_ID')).decode('utf-8')
+        noobaa_secret_key = base64.b64decode(
```
Comment (ebenahar) [inline ocs_ci/ocs/resources/noobaa.py]: same

File: tests/manage/noobaa/test_create_bucket.py [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+    Test creation of a bucket
+    """
+
+    @pytest.mark.skipif(
+        condition=config.ENV_DATA['platform'] != 'AWS',
+        reason="Tests are not running on AWS deployed cluster"
+    )
+    def test_s3_bucket_creation(self, noobaa_obj, buckets):
+        """
+        Test bucket creation using the S3 SDK
```
Comment (vasukulkarni) [inline tests/manage/noobaa/test_create_bucket.py]: There is quite a bit we can utilize from existing ceph resources - will need some minor changes there but worth it

a) https://github.com/red-hat-storage/ceph-qe-scripts/tree/master/rgw/v2/tests/s3_swift
b) https://github.com/ceph/s3-tests

File: ocs_ci/ocs/resources/noobaa.py [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+    _s3_resource = None
+    _ocp_resource = None
+
+    def __init__(self):
+        """
+        Constructor for the NooBaa class
+        """
+        ocp_obj = OCP(kind='noobaa', namespace='noobaa')
+        results = ocp_obj.get()
+        endpoint = (
```
Comment (RazTamir) [inline ocs_ci/ocs/resources/noobaa.py]: do we need this parameter as a class attribute for later usage?
(i.e. use self.endpoint?)

File: ocs_ci/ocs/resources/noobaa.py [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+        results2 = secret_ocp_obj.get(creds_secret_name)
+
+        noobaa_access_key = base64.b64decode(
+            results2.get('data').get('AWS_ACCESS_KEY_ID')
+        ).decode('utf-8')
+        noobaa_secret_key = base64.b64decode(
+            results2.get('data').get('AWS_SECRET_ACCESS_KEY')
+        ).decode('utf-8')
+
+        self._ocp_resource = ocp_obj
```
Comment (RazTamir) [inline ocs_ci/ocs/resources/noobaa.py]: why not assigning this in line 22?

File: ocs_ci/ocs/resources/noobaa.py (hunk 9 above)
Comment (MeridianExplorer) [reply ocs_ci/ocs/resources/noobaa.py]: Not as far as I can tell. AFAIK, the only use we have for the endpoint is for the initial S3 resource creation.

File: ocs_ci/ocs/resources/noobaa.py (hunk 10 above)
Comment (MeridianExplorer) [reply ocs_ci/ocs/resources/noobaa.py]: I feel that assigning all the class members sequentially creates better readability and might save some confusion. Would you recommend assigning it in 22?

Comment (RazTamir) [reply ocs_ci/ocs/resources/noobaa.py]: ack

File: tests/manage/noobaa/test_create_bucket.py (hunk 8 above)
Comment (MeridianExplorer) [reply tests/manage/noobaa/test_create_bucket.py]: That's *very* interesting - thank you!
I shared it with the rest of the NooBaa QE team and I'm sure we'll use it :)

**Comment by MeridianExplorer** [comment]
PR verification log - 
============================= test session starts ==============================
platform linux -- Python 3.7.3, pytest-5.0.1, py-1.8.0, pluggy-0.12.0
rootdir: /home/newuser/repos/ocs-ci-personal-fork, inifile: pytest.ini
plugins: marker-bugzilla-0.9.1.dev2, logger-0.5.0, metadata-1.8.0, html-1.21.1, reportportal-1.0.5
collected 1 item

tests/manage/noobaa/test_create_bucket.py::TestBucketCreation::test_s3_bucket_creation 
-------------------------------- live log setup --------------------------------
13:21:19 - MainThread - tests.conftest - INFO - All logs located at /tmp/ocs-ci-logs-1565259659
13:21:19 - MainThread - ocs_ci.deployment.factory - INFO - Deployment key = aws_ipi
13:21:19 - MainThread - ocs_ci.deployment.factory - INFO - Current deployment platform: AWS,deployment type: ipi
13:21:19 - MainThread - ocs_ci.utility.utils - INFO - Executing command: ./bin/openshift-install version
13:21:19 - MainThread - ocs_ci.utility.utils - INFO - OpenShift Installer version: ./bin/openshift-install v4.2.0-201907300219-dirty
                                                      built from commit 9323138c52b773e31b7c98d8a934f9ddaa6ccaee
                                                      release image registry.svc.ci.openshift.org/ocp/release@sha256:eb93f85c7caec2d7603f384f1a0ae6c004fc3dcdff90c957b074c4ff1b790e2b
13:21:19 - MainThread - ocs_ci.utility.utils - INFO - Executing command: ./bin/oc version
13:21:20 - MainThread - ocs_ci.utility.utils - INFO - OpenShift Client version: Client Version: version.Info{Major:"", Minor:"", GitVersion:"v0.0.0-alpha.0-34-g0d02ccfc", GitCommit:"0d02ccfccbfbabe8be161db4dbb1f42bf00cf7c2", GitTreeState:"clean", BuildDate:"2019-07-28T20:08:21Z", GoVersion:"go1.12.6", Compiler:"gc", Platform:"linux/amd64"}
                                                      Server Version: version.Info{Major:"1", Minor:"14+", GitVersion:"v1.14.0+1682e38", GitCommit:"1682e38", GitTreeState:"clean", BuildDate:"2019-07-28T20:08:40Z", GoVersion:"go1.12.6", Compiler:"gc", Platform:"linux/amd64"}
                                                      OpenShift Version: 4.2.0-0.nightly-2019-07-30-073644
13:21:20 - MainThread - ocs_ci.ocs.openshift_ops - INFO - Testing access to cluster with /home/newuser/repos/beneli-latest-ocs-deployment/auth/kubeconfig
13:21:20 - MainThread - ocs_ci.utility.utils - INFO - Executing command: oc cluster-info
13:21:21 - MainThread - ocs_ci.ocs.openshift_ops - INFO - Access to cluster is OK!
13:21:21 - MainThread - ocs_ci.deployment.aws - WARNING - OCP cluster is already running, skipping installation
13:21:21 - MainThread - ocs_ci.utility.utils - INFO - Executing command: oc -n openshift-storage --kubeconfig /home/newuser/repos/beneli-latest-ocs-deployment/auth/kubeconfig get CephCluster  -o yaml
13:21:22 - MainThread - ocs_ci.deployment.deployment - WARNING - OCS cluster already exists
13:21:22 - MainThread - ocs_ci.utility.utils - INFO - Executing command: oc -n noobaa --kubeconfig /home/newuser/repos/beneli-latest-ocs-deployment/auth/kubeconfig get noobaa  -o yaml
13:21:24 - MainThread - ocs_ci.utility.utils - INFO - Executing command: oc -n noobaa --kubeconfig /home/newuser/repos/beneli-latest-ocs-deployment/auth/kubeconfig get secret noobaa-admin -o yaml
-------------------------------- live log call ---------------------------------
13:21:25 - MainThread - tests.manage.noobaa.test_create_bucket - INFO - Creating new bucket - s3-bucket-testbucketcreation-0813212570
PASSED                                                                   [100%]

=========================== 1 passed in 8.00 seconds ===========================

Process finished with exit code 0

**Comment by vasukulkarni** [comment]
@MeridianExplorer Please squash and merge next time!

**Comment by vavuthu** [comment]
@MeridianExplorer is this test case need separate deployment, because I can see there is a namespace "noobaa" in PR.  In such case, this test fails in our Jenkins.

**Comment by MeridianExplorer** [comment]
@vasukulkarni - I'm so sorry, I had some troubles with internal rebases and conflicts and absolutely forgot. This won't happen again.

@vavuthu - You're correct. Thankfully, @ebenahar pushed an update that skips the test for now.

**Comment by vasukulkarni** [comment]
@MeridianExplorer no issues,  👍 looking forward to many more PRs from you :)

**Pull request merged**